eclair-core/src/main/scala/fr/acinq/eclair/payment/offer/OfferManager.scala

```diff
@@ -0,0 +1,331 @@
+
+    def apply(nodeParams: NodeParams,
+              invoiceRequest: InvoiceRequest,
+              offerHandler: ActorRef[HandleInvoiceRequest],
+              nodeKey: PrivateKey,
+              router: akka.actor.ActorRef,
+              replyPath: OnionMessages.Destination,
+              postman: ActorRef[Postman.SendMessage]): Behavior[Command] = {
+      Behaviors.setup(context => {
+        Behaviors.receiveMessagePartial {
+          case RequestInvoice =>
+            offerHandler ! HandleInvoiceRequest(context.self, invoiceRequest)
+            waitForHandler(nodeParams, invoiceRequest, nodeKey, router, replyPath, postman)
+        }
+      })
+    }
+
+    private def waitForHandler(nodeParams: NodeParams,
+                               invoiceRequest: InvoiceRequest,
+                               nodeKey: PrivateKey,
+                               router: akka.actor.ActorRef,
+                               replyPath: OnionMessages.Destination,
+                               postman: ActorRef[Postman.SendMessage]): Behavior[Command] = {
+      Behaviors.setup(context => {
+        Behaviors.receiveMessagePartial {
+          case RejectRequest =>
+            Behaviors.stopped
+          case ApproveRequest(amount, routes, data, additionalTlvs, customTlvs) =>
+            val preimage = randomBytes32()
+            val metadata = PaymentMetadata(preimage, invoiceRequest.payerId, TimestampSecond.now(), invoiceRequest.quantity, amount, data)
+            val signedMetadata = SignedMetadata(nodeKey, invoiceRequest.offer.offerId, metadata)
+            val pathId = signedMetadataCodec.encode(signedMetadata).require.bytes
+            val receivePayment = MultiPartHandler.ReceiveOfferPayment(nodeKey, invoiceRequest, routes, router, preimage, pathId, additionalTlvs, customTlvs)
+            val child = context.spawnAnonymous(CreateInvoiceActor(nodeParams))
+            child ! CreateInvoiceActor.CreateBolt12Invoice(context.messageAdapter[Try[Bolt12Invoice]](WrappedInvoice), receivePayment)
+            waitForInvoice(replyPath, postman)
+        }
+      })
+    }
+
+    private def waitForInvoice(replyPath: OnionMessages.Destination,
+                               postman: ActorRef[Postman.SendMessage]): Behavior[Command] = {
+      Behaviors.setup(context => {
+        Behaviors.receiveMessagePartial {
+          case WrappedInvoice(Success(invoice)) =>
+            postman ! Postman.SendMessage(Nil, replyPath, None, TlvStream(OnionMessagePayloadTlv.Invoice(invoice.records)), context.messageAdapter[Postman.OnionMessageResponse](WrappedOnionMessageResponse), 0 seconds)
+            waitForSent()
+          case WrappedInvoice(_) =>
+            Behaviors.stopped
+        }
+      })
+    }
+
+    private def waitForSent(): Behavior[Command] = {
+      Behaviors.receiveMessagePartial {
+        case WrappedOnionMessageResponse(_) =>
+          Behaviors.stopped
+      }
+    }
+  }
+
+  object PaymentActor {
```

Comment:
This actor needs to use a timeout after which we automatically reject the payment. We cannot fully trust plugins to be correctly written, and if a plugin doesn't answer and the HTLC times out, this will lead to a channel force-close which isn't acceptable. If the plugin doesn't answer in a timely manner, the safe choice is to fail the HTLC and stop this actor.